### PR TITLE
drivers:lpadc:enable nxp lpadc acquisition time feature

### LINKED ISSUE #68707: Enable acquisition_time for NXP adc_mcux_lpadc.c driver
Currently, the adc_mcux_lpadc.c driver does not support the[ acquisition_time](https://github.com/zephyrproject-rtos/zephyr/blob/129b73ce266e1eabc518333b38d2971c207685fa/drivers/adc/adc_mcux_lpadc.c#L78) configuration.  The HAL does support configuring the [sampleTimeMode ](https://github.com/zephyrproject-rtos/hal_nxp/blob/d45b14c198d778658b7853b48378d2e132a6c4be/mcux/mcux-sdk/drivers/lpadc/fsl_lpadc.c#L460) in the [lpadc_conv_command_config_t struct](https://github.com/zephyrproject-rtos/hal_nxp/blob/d45b14c198d778658b7853b48378d2e132a6c4be/mcux/mcux-sdk/drivers/lpadc/fsl_lpadc.h#L484).

### PR DESCRIPTION
**Pull request by ZhaoxiangJin**
Enable nxp lpadc acquisition time feature, acquisition time can be expressed in microseconds, nanosecond, and ADC tickets, this patch has been processed these three different situations. When the input acquisition time unit is ADC tickets, it only needs to be matched with the elements in the sample time array. If the input acquisition time unit is nanoseconds or microseconds, first convert the unit of the elements in the sample time array from ADCK to nanoseconds, and then use the dichotomy to match the option closest to the acquisition time from the sample time array.

### CODE REVIEW

**Comment by decsny** [comment]
please read https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations

this commit needs to be split up to multiple before it will be reviewed, see other PRs for examples

**Comment by ZhaoxiangJin** [comment]
> please read https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations
> 
> this commit needs to be split up to multiple before it will be reviewed, see other PRs for examples

**Comment by ZhaoxiangJin** [comment]
> please read https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#contributor-expectations
> 
> this commit needs to be split up to multiple before it will be reviewed, see other PRs for examples

Thanks for your reminder, I updated the PR and split it into four commits, but the PR was accidentally closed by me. Can you please help me reopen it? Thanks.

**Comment by ZhaoxiangJin** [comment]
could anyone help to review this PR? Thanks a lot.

File: samples/drivers/adc/boards/lpcxpresso55s36.overlay [hunk 1]
```diff
@@ -26,7 +26,7 @@
 		reg = <0>;
 		zephyr,gain = "ADC_GAIN_1";
 		zephyr,reference = "ADC_REF_EXTERNAL0";
-		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
+		zephyr,acquisition-time = <ADC_ACQ_TIME(ADC_ACQ_TIME_TICKS, 7)>;
```
Comment (decsny) [inline samples/drivers/adc/boards/lpcxpresso55s36.overlay:29]: is there a reason for 7 in particular?

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2022 NXP
+ * Copyright 2022,2024 NXP
```
Comment (decsny) [inline dts/arm/nxp/nxp_lpc55S3x_common.dtsi:2]: need space after comma

File: dts/arm/nxp/nxp_rt6xx_common.dtsi [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, NXP
+ * Copyright (c) 2020,2024 NXP
```
Comment (decsny) [inline dts/arm/nxp/nxp_rt6xx_common.dtsi:2]: space after comma and remove the (c)

File: drivers/clock_control/clock_control_mcux_ccm.c [hunk 4]
```diff
@@ -377,6 +377,15 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
 					/ (CLOCK_GetDiv(kCLOCK_Flexio2Div) + 1);
 	} break;
 #endif
+
+#ifdef CONFIG_ADC_MCUX_LPADC
+	case IMX_CCM_LPADC1_CLK:
+		*rate = CLOCK_GetRootClockFreq(kCLOCK_Root_Adc1);
+		break;
+	case IMX_CCM_LPADC2_CLK:
+		*rate = CLOCK_GetRootClockFreq(kCLOCK_Root_Adc2);
```
Comment (decsny) [inline drivers/clock_control/clock_control_mcux_ccm.c:386]: do all platforms have these two enums defined ? ie what if a platform has only 1 adc, will it cause a build failure, or do we have no cases like that yet (if not, then it's fine for now)

File: drivers/adc/adc_mcux_lpadc.c [hunk 5]
```diff
@@ -93,6 +190,12 @@ static int mcux_lpadc_channel_setup(const struct device *dev,
 
 	LPADC_GetDefaultConvCommandConfig(cmd);
 
+	/* Configure LPADC acquisition time. */
+	if (mcux_lpadc_acquisition_time_setup(dev, channel_cfg->acquisition_time, cmd)) {
+		LOG_ERR(" LPADC acquisition time setting failed");
```
Comment (decsny) [inline drivers/adc/adc_mcux_lpadc.c:195]: extra space in the beginning not needed

File: drivers/adc/adc_mcux_lpadc.c [hunk 6]
```diff
@@ -57,7 +59,107 @@ struct mcux_lpadc_data {
...
 };
 
+static int mcux_lpadc_acquisition_time_setup(const struct device *dev, uint16_t acq_time,
+					     lpadc_conv_command_config_t *cmd)
+{
+	const struct mcux_lpadc_config *config = dev->config;
+	uint32_t adc_freq_hz = 0;
+	float adc_cycle_ns = 0.0;
+	/* The unit of elements in the sample time array is ADCK. */
+	float sample_time_array[] = {3.0, 5.0, 7.0, 11.0, 19.0, 35.0, 67.0, 131.0};
```
Comment (decsny) [inline drivers/adc/adc_mcux_lpadc.c:69]: why is it needed to use floating points here ?

**Review by decsny — DISMISSED** [review]
The commits are well structured except for the fact that they are in a backwards order. You need to put the clock control commit first, then DTS changes, then the driver changes, then the sample overlays. This I think will ensure bisectibility of the commit history.

I still have to review the mcux_lpadc_acquisition_time_setup function more closely but first help me to understand why you are using floating point arithmetic so I have some better background about the code. Since floating point arithmetic will incur a cost on the platforms without an FPU especially there must be a reason?

**Comment by decsny** [comment]
@yvanderv please review the copyright changes of files that did not have NXP copyright, I don't know the rule here

File: samples/drivers/adc/boards/lpcxpresso55s36.overlay (hunk 1 above)
Comment (ZhaoxiangJin) [reply samples/drivers/adc/boards/lpcxpresso55s36.overlay:29]: For the ADC, the longer the sampling time is set, the higher the accuracy of the data obtained, but too high a sampling time will cause the sampling rate to decrease, so I chose a more compromised configuration option.

File: dts/arm/nxp/nxp_lpc55S3x_common.dtsi (hunk 2 above)
Comment (ZhaoxiangJin) [reply dts/arm/nxp/nxp_lpc55S3x_common.dtsi:2]: thanks for your reminder, I will update it.

File: dts/arm/nxp/nxp_rt6xx_common.dtsi (hunk 3 above)
Comment (ZhaoxiangJin) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:2]: thanks for your reminder, I will update it.

File: drivers/clock_control/clock_control_mcux_ccm.c (hunk 4 above)
Comment (ZhaoxiangJin) [reply drivers/clock_control/clock_control_mcux_ccm.c:386]: Currently, for all platforms equipped with LPADC and CCM controllers(RT116x, RT117x), the number of ADC instances is 2.

File: drivers/adc/adc_mcux_lpadc.c (hunk 5 above)
Comment (ZhaoxiangJin) [reply drivers/adc/adc_mcux_lpadc.c:195]: thanks for your reminder, I will remove the space.

File: drivers/adc/adc_mcux_lpadc.c (hunk 6 above)
Comment (ZhaoxiangJin) [reply drivers/adc/adc_mcux_lpadc.c:69]: In this patch logic, I convert the sample time array elements unit from ADCK to ns, that is each element needs to be multiplied by the variable  **adc_cycle_ns**, and the variable adc_cycle_ns is float point type.

**Comment by ZhaoxiangJin** [comment]
> The commits are well structured except for the fact that they are in a backwards order. You need to put the clock control commit first, then DTS changes, then the driver changes, then the sample overlays. This I think will ensure bisectibility of the commit history.
> 
> I still have to review the mcux_lpadc_acquisition_time_setup function more closely but first help me to understand why you are using floating point arithmetic so I have some better background about the code. Since floating point arithmetic will incur a cost on the platforms without an FPU especially there must be a reason?

Sorry for the late commit sequence update. The reason for using the float point type is the float point value are introduced when calculating the ADC period based on the ADC clock.

**Comment by decsny** [comment]
I am still confused why does adc_cycle_ns even need to be a float? It is being used to calculate the float sample_time_array values, but then these values are later casted to an int when used. Can't all of these be long ints perhaps? Or am I missing something

**Comment by ZhaoxiangJin** [comment]
> I am still confused why does adc_cycle_ns even need to be a float? It is being used to calculate the float sample_time_array values, but then these values are later casted to an int when used. Can't all of these be integer types perhaps? Or am I missing something

Why I think the variable **adc_cycle_ns** needs to be set to a floating point type, I would like to give an example to illustrate. Under NXP LPC55S36 platform, the default ADC clock frequency is 48MHz, so the ADC cycle is approximately 20.833 ns (if variable **adc_cycle_ns** is set to an integer type, then this number is 20). 

case1(float point type, 20.833):
sample_time_array[0] = 20.833 * 3 =  62.499 ns
sample_time_array[1] = 20.833 * 5 = 104.165 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option **kLPADC_SampleTimeADCK3** will be set to the corresponding bitfield.

case1(integer type, 20):
sample_time_array[0] = 20 * 3 =  60 ns
sample_time_array[1] = 20 * 5 = 100 ns
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option **kLPADC_SampleTimeADCK5** will be set to the corresponding bitfield. 

If we use integer type in the code, the accuracy of the calculation results decreases. At this clock frequency (48MHz), and application configuration options (82ns), the optimal configuration item should be  **kLPADC_SampleTimeADCK3**.

Currently, my local code has changed the ADC sample time array (**sample_time_array**) to the integer type. The reason is that converting the final calculation result into an integer will not cause too much loss of accuracy, and at the same time reduces the calculation pressure of chips without FPU platforms (the subsequent calculation process uses array elements to compare with the **acquisition_time_value**).

for example
sample_time_array[0] = 20.833 * 3 =  62.499 ns = 62
sample_time_array[1] = 20.833 * 5 = 104.165 ns = 104
if the parameter ADC acquisition time passed by the application layer is 82ns, then the option **kLPADC_SampleTimeADCK3** will be set to the corresponding bitfield.

**Comment by ZhaoxiangJin** [comment]
Recreate a new PR (https://github.com/zephyrproject-rtos/zephyr/pull/70636) based on the nxp-upstream repository.

**Pull request closed without merging**